### PR TITLE
Add content verification to the plugin installation

### PR DESCRIPTION
**Pull request by rogaha**
- remove the `tgz` artifacts and use `docker plugin install` instead

Signed-off-by: Roberto Gandolfo Hashioka <roberto_hashioka@hotmail.com>

### CODE REVIEW

File: rpm/Makefile [hunk 1]
```diff
@@ -1,9 +1,8 @@
 ARCH=$(shell uname -m)
 ALPINE_IMG:=$(shell $(CURDIR)/../detect_alpine_image)
-KERNEL:=$(shell docker version --format '{{.Server.Os}}')
+KERNEL:=linux
```
Comment (rogaha) [inline rpm/Makefile]: 👍 I'll do that.

File: rpm/Makefile [hunk 2]
```diff
@@ -1,9 +1,8 @@
...
 ENGINE_DIR:=$(CURDIR)/../../engine
 COMMON_DIR:=$(CURDIR)/../common
```
Comment (rogaha) [inline rpm/Makefile]: that's correct. We are only using it as a `volume` in order to access the `telemetry script`.

File: rpm/centos-7/docker-ce.spec [hunk 3]
```diff
@@ -70,7 +69,7 @@ TMP_GOPATH="/go" hack/dockerfile/install-binaries.sh runc-dynamic containerd-dyn
 hack/make.sh dynbinary
 popd
 mkdir -p plugin
-printf '{"edition_type":"ce","edition_name":"%s","edition_version":"%s"}\n' "${DISTRO}" "%{_version}" > plugin/.plugin-metadata
+printf '{"edition_type":"ce","edition_name":"%s","edition_version":"%s", "plugin_version":"%s"}\n' "${DISTRO}" "%{_version}" "%{_plugin_version}" > plugin/.plugin-metadata
```
Comment (rogaha) [inline rpm/centos-7/docker-ce.spec]: 👍 agreed. I'll rename it.

File: common/load-telemetry-plugin [hunk 4]
```diff
@@ -16,7 +15,14 @@ if [ -z ${UUID+x} ]; then
     UUID=$(cat /proc/sys/kernel/random/uuid | tee $BASE/engine_uuid)
 fi
 
+# check if the file exists
```
Comment (rogaha) [inline common/load-telemetry-plugin]: 👍

File: common/load-telemetry-plugin [hunk 5]
```diff
@@ -16,7 +15,14 @@ if [ -z ${UUID+x} ]; then
...
+if ! [ -e "$PLUGIN_METADATA" ]; then
+        # plugin metadata not found, exiting...
```
Comment (rogaha) [inline common/load-telemetry-plugin]: 👍

**Comment by rogaha** [comment]
hi @thaJeztah thanks a lot for your feedback here, the `plugin` installation will be much more robust thanks to your comments.

>It changes the installation
>The /var/lib/docker path is hard-coded, and this path will now always be created. In the current (17.06 and before) installation, /var/lib/docker is created by the docker daemon when started, and this path is configurable. You can try this by installing docker on an RPM based install (which does not automatically start the service after creation). After installation the /var/lib/docker directory is not present.

This directory was being used only to keep the `.package-information` (a.k.a `.plugin_metadata`), I'll remove this requirement and use the `docker` API to get the the location of this file. 

>Perhaps the script itself should be somewhere in /usr/bin/docker-telemetrics-install (or whatever)

The script is located in `/usr/bin`.

>Has this been tested with (e.g.) SELinux?
I'll remove this requirement.

>For systemd; the plugin should have it's own unit-file with After=docker.service (or similar). This also allows users to disable the service separate from disabling Docker (without having to either dive into systemd drop-in files, or (not the correct way, but very likely to happen), editing the main unit tile).

@thaJeztah users can easily disable the `script` by using the `docker CLI`. We can create it's own `unit-file`, but it will increase the `complexity` of this change as the `load` script will need to support `start/stop/restart`. I would prefer to keep it simple for now. 

>This plugin is new; bugs may be present, improvements may be needed. Including it in the same package as the Docker engine forces us to issue a new release of Docker in case a bug has to be fixed. For the Desktop products (Docker for Mac / Docker for Windows), automated updates are possible, and would allow pushing an updated version between releases. For other platforms, this is more troublesome.

@thaJeztah we are just loading a `docker plugin` by default using the `own` `Docker` APIs, we don't have time to make this change now just to make it more `segregated`. Let's keep the changes as small as possible while addressing your concerns. 

I'll submit a path to address (1), (2) and maybe (3) if we all agree that keeping the `load plugin script` in a separate `unit` is work the extra `management` required changes.

**Comment by rogaha** [comment]
>The /var/lib/docker path is hard-coded, and this path will now always be created. In the current (17.06 and before) installation, /var/lib/docker is created by the docker daemon when started, and this path is configurable. You can try this by installing docker on an RPM based install (which does not automatically start the service after creation). After installation the /var/lib/docker directory is not present.

@thaJeztah the `/var/lib/docker` directory will exist after the installation, it's created by the `installation` script. See it in `override_dh_auto_install` section.

**Comment by rogaha** [comment]
@thaJeztah PTAL

**Comment by rogaha** [comment]
@thaJeztah I really appreciate your comments here. Having the `/var/lib/docker` directory created, doesn’t impact the `location` of the docker installation, users can still change the `Docker Root Dir`, but we will use the `/var/lib/docker` to keep the state of the `edition-information` file. Do you see any concerns with that?

The RPM and DEB installations already create the `/usr/lib/docker` directory at the installation time, but I think /var/lib/docker` is the proper location to keep the `docker package/edition` metadata.

**Comment by rogaha** [comment]
if you still see any `concerns` about the way we are loading the `plugin` please let me know. I would like to address all your comments by EOD.

**Comment by thaJeztah** [comment]
>  the `/var/lib/docker` directory will exist after the installation, it's created by the installation script. See it in `override_dh_auto_install` section.

Yes, so this was also introduced by the metrics install changes https://github.com/docker/docker-ce-packaging/blame/f608c8c590676f30bc1fe05bf65e6716d3e97a3a/deb/common/rules#L36-L39, but not the case in the current ("live") packages for 17.06

File: common/docker-telemetry-plugin [hunk 6]
```diff
@@ -27,40 +16,53 @@ if [ -n "$DOCKER_SOCKET" ]; then
...
+if [ -z ${UUID+x} ]; then
+    UUID=$(cat /proc/sys/kernel/random/uuid | tee $BASE/engine_uuid)
 fi
 
 # store the metrics plugin state
-plugin_name=$(docker plugin ls | grep docker/telemetry | awk 'NR==1{print $2}')
-plugin_metadata=$TAR_DIR/.plugin-metadata
+plugin_name=$(docker plugin ls | grep $PLUGIN_NAME | awk 'NR==1{print $2}')
 if docker inspect $plugin_name > /dev/null 2>&1; then
         plugin_state=$(docker plugin inspect $plugin_name | grep -o '"Enabled":.*' | sed 's/,/\\\}/g' | sed 's/ //g')
```
Comment (thaJeztah) [inline common/docker-telemetry-plugin]: To get just the `true` / `false` for `Enabled`, consider using `docker plugin inspect -f '{{.Enabled}}' $plugin_name`

Comment (rogaha) [inline common/docker-telemetry-plugin]: thanks for the feedback @thaJeztah. I'll use that instead!

**Comment by rogaha** [comment]
@thaJeztah PTAL

**Comment by rogaha** [comment]
closing tis `PR` as we agreed on using `checksum` against the `tgz` file instead.

**Pull request closed without merging**